core/src/main/java/org/jruby/ir/interpreter/InterpreterEngine.java

```diff
@@ -433,7 +433,7 @@ protected static IRubyObject processReturnOp(ThreadContext context, Block block,
                 // This assumes that scopes with break instr. have a frame / dynamic scope
                 // pushed so that we can get to its static scope. For-loops now always have
                 // a dyn-scope pushed onto stack which makes this work in all scenarios.
-                return IRRuntimeHelpers.initiateBreak(context, currDynScope, rv, blockType);
+                return IRRuntimeHelpers.initiateBreak(context, currDynScope, rv, block, blockType);
```

Comment:
Turns out blockType was only used by the two cases that always have a non-null block (as we discussed on IRC, non-local return and break *should* only be emitted for block bodies) so I just removed that null check and made each method look at block.type directly. Addresses this and other comment.